drivers/ethernet/eth_xlnx_gem.c

```diff
@@ -0,0 +1,1724 @@
+	if (k_delayed_work_remaining_get(
+		&dev_data->phy_poll_delayed_work) == 0) {
+		k_delayed_work_submit(&dev_data->phy_poll_delayed_work,
+			K_NO_WAIT);
+	}
+
+	LOG_DBG("%s started", dev->name);
+	return 0;
+}
+
+/**
+ * @brief GEM device stop function
+ * GEM device stop function. Disables all interrupts, disables
+ * RX and TX, clears all status registers. If no PHY is managed
+ * by the current driver instance, this function also declares
+ * the physical link down.
+ *
+ * @param dev Pointer to the device data
+ * @retval    0 upon successful completion
+ */
+static int eth_xlnx_gem_stop_device(const struct device *dev)
+{
+	const struct eth_xlnx_gem_dev_cfg *dev_conf = DEV_CFG(dev);
+	struct eth_xlnx_gem_dev_data *dev_data = DEV_DATA(dev);
+	uint32_t reg_val = 0;
+
+	if (dev_data->started == 0) {
+		return 0;
+	}
+	dev_data->started = 0;
+
+	/* Cancel the delayed work that polls the link state */
+	if (k_delayed_work_remaining_get(
+		&dev_data->phy_poll_delayed_work) != 0) {
+		k_delayed_work_cancel(&dev_data->phy_poll_delayed_work);
+	}
+
+	/* RX and TX disable */
+	reg_val  = sys_read32(
+		dev_conf->base_addr + ETH_XLNX_GEM_NWCTRL_OFFSET);
+	reg_val &= (~(ETH_XLNX_GEM_NWCTRL_RXEN_BIT |
+		ETH_XLNX_GEM_NWCTRL_TXEN_BIT));
+	sys_write32(reg_val,
+		dev_conf->base_addr + ETH_XLNX_GEM_NWCTRL_OFFSET);
+
+	/* Disable & clear all the MAC interrupts */
+	sys_write32(ETH_XLNX_GEM_IXR_ALL_MASK,
+		dev_conf->base_addr + ETH_XLNX_GEM_IDR_OFFSET);
+	sys_write32(ETH_XLNX_GEM_IXR_ALL_MASK,
+		dev_conf->base_addr + ETH_XLNX_GEM_ISR_OFFSET);
+
+	/* Clear RX & TX status registers */
+	sys_write32(0xFFFFFFFF,
+		dev_conf->base_addr + ETH_XLNX_GEM_TXSR_OFFSET);
+	sys_write32(0xFFFFFFFF,
+		dev_conf->base_addr + ETH_XLNX_GEM_RXSR_OFFSET);
+
+	LOG_DBG("%s stopped", dev->name);
+	return 0;
+}
+
+/**
+ * @brief GEM capability request function
+ * Returns the capabilities of the GEM controller as an enumeration.
+ * All of the data returned is derived from the device configuration
+ * of the current GEM device instance.
+ *
+ * @param dev Pointer to the device data
+ * @return Enumeration containing the current GEM device's capabilities
+ */
+static enum ethernet_hw_caps eth_xlnx_gem_get_capabilities(
+	const struct device *dev)
+{
+	const struct eth_xlnx_gem_dev_cfg *dev_conf = DEV_CFG(dev);
+	enum ethernet_hw_caps caps = (enum ethernet_hw_caps)0;
+
+	if (dev_conf->max_link_speed == LINK_1GBIT) {
+		if (dev_conf->phy_advertise_lower == 1) {
+			caps |= (
+				  ETHERNET_LINK_1000BASE_T
+				| ETHERNET_LINK_100BASE_T
+				| ETHERNET_LINK_10BASE_T);
+		} else {
+			caps |= ETHERNET_LINK_1000BASE_T;
+		}
+	} else if (dev_conf->max_link_speed == LINK_100MBIT) {
+		if (dev_conf->phy_advertise_lower == 1) {
+			caps |= (
+				  ETHERNET_LINK_100BASE_T
+				| ETHERNET_LINK_10BASE_T);
+		} else {
+			caps |= ETHERNET_LINK_100BASE_T;
+		}
+	} else {
+		caps |= ETHERNET_LINK_10BASE_T;
+	}
+
+	if (dev_conf->enable_rx_chksum_offload == 1) {
```

Comment:
These vars, that are "iffed" here and below, seem to be boolean by nature. You could even combine all the booleans in dev_conf to bit fields to save some space.